File: packages/python/server/starfish_server/encryption/encrypted_store.py
"""AES-256-GCM encrypted object store wrapper."""


import base64
import os

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from starfish_protocol.crypto import _derive_key, IV_BYTES
from starfish_server.storage.base import AbstractObjectStore, StoreContext
from starfish_server.constants import HKDF_INFO_DEFAULT


class EncryptedObjectStore(AbstractObjectStore):
    """Wraps an AbstractObjectStore to transparently encrypt/decrypt all values.

    Keys (paths) are NOT encrypted — only the stored content.
    """

    def __init__(
        self,
        inner: AbstractObjectStore,
        secret: str,
        salt: str,
        info: str = HKDF_INFO_DEFAULT,
    ) -> None:
        self._inner = inner
        key = _derive_key(secret, salt, info.encode("utf-8"))
        self._aesgcm = AESGCM(key)

    def _encrypt(self, plaintext: str) -> str:
        iv = os.urandom(IV_BYTES)
        data = plaintext.encode("utf-8")
        ciphertext = self._aesgcm.encrypt(iv, data, None)
        combined = iv + ciphertext
        return base64.b64encode(combined).decode("ascii")

    def _decrypt(self, encoded: str) -> str:
        combined = base64.b64decode(encoded)
        if len(combined) < IV_BYTES:
            raise ValueError("Encrypted data is too short")
        iv = combined[:IV_BYTES]
        ciphertext = combined[IV_BYTES:]
        try:
            plaintext = self._aesgcm.decrypt(iv, ciphertext, None)
        except InvalidTag as exc:
            raise ValueError("Decryption failed: data may be tampered or key is incorrect") from exc
        return plaintext.decode("utf-8")

    async def get_string(self, key: str, *, context: StoreContext | None = None) -> str | None:
        raw = await self._inner.get_string(key, context=context)
        if raw is None:
            return None
        return self._decrypt(raw)

    async def put(
        self,
        key: str,
        body: str,
        *,
        content_type: str | None = None,
        cache_control: str | None = None,
        context: StoreContext | None = None,
    ) -> None:
        encrypted = self._encrypt(body)
        await self._inner.put(key, encrypted, content_type=content_type, cache_control=cache_control, context=context)

    async def list_keys(
        self,
        prefix: str,
        *,
        start_after: str | None = None,
        limit: int | None = None,
        context: StoreContext | None = None,
    ) -> list[str]:
        return await self._inner.list_keys(prefix, start_after=start_after, limit=limit, context=context)

    async def delete(self, key: str, *, context: StoreContext | None = None) -> None:
        await self._inner.delete(key, context=context)

    async def delete_many(self, keys: list[str], *, context: StoreContext | None = None) -> None:
        await self._inner.delete_many(keys, context=context)

File: packages/ts/server/src/encryption/encrypted_store.ts
import { deriveKey, getCrypto, getBase64, IV_BYTES } from "@starfish/protocol"
import { AbstractObjectStore } from "../storage/base.js"
import { HKDF_INFO_DEFAULT } from "../constants.js"

const ALGO = "AES-GCM"

export class EncryptedObjectStore extends AbstractObjectStore {
  private readonly inner: AbstractObjectStore
  private readonly keyPromise: Promise<CryptoKey>

  constructor(
    inner: AbstractObjectStore,
    secret: string,
    salt: string,
    info = HKDF_INFO_DEFAULT,
  ) {
    super()
    this.inner = inner
    this.keyPromise = deriveKey(secret, salt, info)
  }

  private async encrypt(plaintext: string): Promise<string> {
    const key = await this.keyPromise
    const c = getCrypto()
    const b64 = getBase64()
    const data = new TextEncoder().encode(plaintext)
    const iv = c.getRandomValues(new Uint8Array(IV_BYTES))
    const ciphertext = await c.subtle.encrypt({ name: ALGO, iv }, key, data)
    const combined = new Uint8Array(iv.length + ciphertext.byteLength)
    combined.set(iv)
    combined.set(new Uint8Array(ciphertext), iv.length)
    return b64.encode(combined)
  }

  private async decrypt(encoded: string): Promise<string> {
    const key = await this.keyPromise
    const c = getCrypto()
    const b64 = getBase64()
    const combined = b64.decode(encoded)
    if (combined.length < IV_BYTES) {
      throw new Error("Encrypted data is too short")
    }
    const iv = combined.slice(0, IV_BYTES)
    const ciphertext = combined.slice(IV_BYTES)
    const plaintext = await c.subtle.decrypt({ name: ALGO, iv }, key, ciphertext)
    return new TextDecoder().decode(plaintext)
  }

  async getString(key: string): Promise<string | null> {
    const raw = await this.inner.getString(key)
    if (raw === null) return null
    return this.decrypt(raw)
  }

  async put(
    key: string,
    body: string,
    options?: { contentType?: string; cacheControl?: string },
  ): Promise<void> {
    const encrypted = await this.encrypt(body)
    return this.inner.put(key, encrypted, options)
  }

  async listKeys(
    prefix: string,
    options?: { startAfter?: string; limit?: number },
  ): Promise<string[]> {
    return this.inner.listKeys(prefix, options)
  }

  async delete(key: string): Promise<void> {
    return this.inner.delete(key)
  }

  async deleteMany(keys: string[]): Promise<void> {
    return this.inner.deleteMany(keys)
  }
}
